Better support for non-deterministic external-names by updating critical annotations

twobiers opened the pull request:
### Description of your changes

This PR ensures that critical annotations are stored when a resource is created. Some resources have non-deterministic external-names, which are never updated when the `mangementPolicies` don't contain `LateInitialize`.
So far, there is an implicit contract that an Observation updating the external-name will be eventually stored as part of the LateInitialize process. However, that should only affect updates to the `spec` and not the annotations like external-name.
More Context can be found here: https://github.com/crossplane/crossplane/issues/5918

I'm not sure how a unit test can be expressed, as I'm not really familiar with the setup. If the change is approved, I think it would make sense to backport it aswell.

Fixes:
- https://github.com/crossplane/crossplane/issues/5918
- https://github.com/crossplane-contrib/provider-upjet-azuread/issues/159
- https://github.com/crossplane-contrib/provider-upjet-aws/issues/1567
- https://github.com/crossplane-contrib/provider-upjet-azure/issues/802

Maybe fixes aswell:
- https://github.com/crossplane-contrib/provider-upjet-aws/issues/1482

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~ 
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

twobiers committed:
Update critical annotations when a resource exists

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
Ensure only annotations are updated

As the UpdateCriticialAnnotations function is now not exclusively called in the creation process, we have to ensure no other fields like the spec are updated, so we don't interfer with the normal LateInitialize logic

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
Update test assumption to update annotations

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
Add more details to comment

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers commented:
@jeanduplessis Thanks for your kind comment. I'm able to work on this and have rebased the branch onto master.

twobiers committed:
Add MockPatch to modern test case

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
Add all missing mocks

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

erhancagirici commented on pkg/reconciler/managed/api.go:
```diff
@@ -277,15 +277,25 @@ func NewRetryingCriticalAnnotationUpdater(c client.Client) *RetryingCriticalAnno
+			return err
+		}
+
+		err = u.client.Patch(ctx, o, client.RawPatch(types.MergePatchType, patchData), client.FieldOwner(fieldOwnerAPISimpleRefResolver), client.ForceOwnership)
```
Just wondering how does this interact with the field ownerships of annotations that XP does not manage (like a custom annotation set by a user or some other controllers, tooling etc), and whether it is a thing to worry about regarding server-side apply etc.
Would we cause a race regarding field ownerships?

ulucinar commented on pkg/reconciler/managed/api.go:
```diff
@@ -277,15 +277,25 @@ func NewRetryingCriticalAnnotationUpdater(c client.Client) *RetryingCriticalAnno
 	err := retry.OnError(retry.DefaultRetry, func(err error) bool {
 		return !errors.Is(err, context.Canceled)
 	}, func() error {
-		err := u.client.Update(ctx, o)
```
I like this update operation is being replaced by SSA.

erhancagirici reviewed (commented):
Thanks for the PR! I've left some comments regarding the mechanics, otherwise LGTM conceptually. As you mentioned in the doc comments, this will be sort of a band-aid until a more async-aware implementation.

ulucinar reviewed (commented):
Thanks @twobiers for working on this issue. Left some comments for us to discuss.

negz commented on pkg/reconciler/managed/reconciler.go:
```diff
@@ -1408,6 +1408,21 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
+		// after the creation, but later as part of an asynchronous process.
+		// When Crossplane supports asynchronous creation of resources natively, this logic
+		// might not be needed anymore and can be revisited. 
+		if err := r.managed.UpdateCriticalAnnotations(ctx, managed); err != nil {
```
Perhaps we should make writing these annotations optional. I'd suggest making them opt-out since that's the safest path. The option would be specified by the provider author - so they can disable these annotations if they know for sure they're not needed (i.e. naming is deterministic, API is strongly consistent). I remember discussing this with someone recently, but can't find a tracking issue.

(If we do make them optional, I think we could do it pretty easily by injecting a no-op implementation of the annotation updater.)

twobiers committed:
Add critical annotations handling in UpdateCriticalAnnotations method

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers commented:
I just had the idea that a function in the pipeline could also ensure those annotations are persisted. It would not be an ideal solution, but at least a possible hotfix while async is not fully supported.

jeanduplessis commented:
@bobh66 this PR now needs a review and approval from a @crossplane/crossplane-maintainers to move it forward.

twobiers commented:
/fresh

jwefers commented:
Does this fix https://github.com/crossplane/upjet/issues/632 ?

bobh66 commented:
@twobiers can you check the lint and unit-test errors?  I'd like to get this moving if there are no other solutions in the works, and I don't think there are.

twobiers committed:
Set a critical annotation

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
apply formatting

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers committed:
fix test assertion

Signed-off-by: twobiers <22715034+twobiers@users.noreply.github.com>

twobiers commented:
@bobh66 I've rebased the branch and fixed the issues in the checks. They do pass locally now. Not sure if you have to kick off the workflows here again manually.
Let me know if you have any further comments on the change
